cpp/include/raft/core/operators.hpp

```diff
@@ -130,10 +130,10 @@ struct div_op {
 };
 
 struct div_checkzero_op {
-  template <typename Type>
-  constexpr RAFT_INLINE_FUNCTION auto operator()(const Type& a, const Type& b) const
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const T1& a, const T2& b) const
```

Comment:
I think, it would be healthier to use the `map_args_op{div_checkzero_op{}, identity_op{}, cast_op<T1>{}}` on the call site rather than changing the semantics here.

Reply:
Oh, wait, I see many of the other binary operators are heterogeneous in their argument types as well. Unless we don't change all of them to use the same type for both arguments, it does not make much sense to hold off this change as well.
Maybe add the explicit type declaration `-> T1` to them, so that the users wouldn't need to try hard to remember the C++ arithmetics type inference rules?

Reply:
I prefer not to cast explicitly if you wouldn't when writing the operation in plain C++. And I think most people wouldn't cast an integer divisor to float when the dividend is float.

Regarding the explicit type declaration, I disagree, it can be either `T1` or `T2` based on the types. E.g if one is float and the other integer, the result is always float. Hence my writing `return T1{0} / T2{1};` to avoid making any assumption.